src/host/readDataCooked.cpp

```diff
@@ -797,453 +817,477 @@ void COOKED_READ_DATA::_transitionState(State state) noexcept
+            for (size_t beg = 0; beg < segment.size();)
             {
-                distance += _writeCharsUnprocessed({ it, nextControlChar });
+                if (res.column >= size.width)
+                {
+                    lines.emplace_back();
+                }
+
+                auto& line = lines.back();
+                res = _layoutLine(line.text, segment, beg, line.columns, size.width);
+                line.columns = res.column;
+
+                if (!dirty)
+                {
+                    line.dirtyBegOffset = line.text.size();
+                    line.dirtyBegColumn = res.column;
+                }
+
+                beg = res.offset;
             }
-            it = nextControlChar;
-        }
-        if (nextControlChar == end)
-        {
-            break;
-        }
 
-        wchar_t buf[2];
-        size_t len = 0;
+            // If this segment ended at the cursor offset, we got our cursor position in rows/columns.
+            if (offsets[i + 1] == _bufferCursor)
+            {
+                cursorPositionFinal = { res.column, gsl::narrow_cast<til::CoordType>(lines.size() - 1) };
+            }
+        }
 
-        const auto wch = *it;
-        if (wch == UNICODE_TAB)
+        pagerPromptEnd = { res.column, gsl::narrow_cast<til::CoordType>(lines.size() - 1) };
+        if (pagerPromptEnd.x >= size.width)
         {
-            buf[0] = L'\t';
-            len = 1;
+            pagerPromptEnd.x = 0;
+            pagerPromptEnd.y++;
         }
-        else
+
+        // If the content got a little shorter than it was before, we need to erase the tail end.
+        // If the last character on a line got removed, we'll skip this code because `remaining`
+        // will be negative, and instead we'll erase it later when we append "  \r" to the lines.
+        // If entire lines got removed, then we'll fix this later when comparing against _pagerContentEnd.y.
+        if (pagerPromptEnd.y <= _pagerPromptEnd.y)
         {
-            // In the interactive mode we replace C0 control characters (0x00-0x1f) with ASCII representations like ^C (= 0x03).
-            buf[0] = L'^';
-            buf[1] = gsl::narrow_cast<wchar_t>(wch + L'@');
-            len = 2;
+            const auto endX = _pagerPromptEnd.y == pagerPromptEnd.y ? _pagerPromptEnd.x : size.width;
+            const auto remaining = endX - pagerPromptEnd.x;
+
+            if (remaining > 0)
+            {
+                auto& line = lines.back();
+
+                // CSI K may be expensive, so use spaces if we can.
+                if (remaining <= 8)
+                {
+                    line.text.append(remaining, L' ');
+                }
+                else
+                {
+                    line.text.append(L"\x1b[K");
+                }
+            }
         }
 
-        if (measureOnly)
+        // Render the popups, if there are any.
+        if (!_popups.empty())

```

Comment:
I'm not really a CMD user so I'm not really familiar with how to get popups to appear. BUT if you tell me how to get popups to appear, I'd be happy to test this out manually to get some extra coverage :)

Reply:
There are 3 "question" popups with F2, F4, and F9, and there's one interactive popup via F7. When you're in the F7 prompt, you can press F9 to select one of the entries immediately. You can read about what the 4 popups are supposed to do here (they're in the order F2, F4, F9, F7): https://github.com/microsoft/terminal/blob/dev/lhecker/14000-vt-cooked-read/src/host/readDataCooked.hpp#L54-L73

I should mention the F-keys that open them in that comment. 🤔

Reply:
Alright, I did a little bit of testing on the popups. Tested against v1.20.11781. Here's a short report of my findings:

## F2
- The UI does look different between the two. I'm guessing that's expected (and I'm guessing it's ok for us to do this). Curious, why though? I'm guessing it's a limitation on switching over to just using VT?
![UI comparison](https://github.com/microsoft/terminal/assets/11050425/90e340d0-ad1b-470b-b2bf-0dda452a7ffc)
- 🐛BUG: 
   - run `echo hello world`
   - F2->`w`
   - expected: `echo hello `
   - actual: the characters are copied over to the input buffer, but they aren't rendered
- ✅ the input buffer is properly updated regardless of the initial content of the input buffer
   - i.e. start with `abcdefghijklmnop` then use `F2` --> input buffer is updated appropriately (data-wise)

## F4
- ✅ input buffer properly updated
   - test: `abcdefghijklmnopqrstuvwxyz` (cursor just after the `f`) + `F4` + `x` --> `abcdefxyz`

## F9
- ✅ correctly retrieves the right command in the history and throws it in the input buffer
   - test: fill history with `echo 1` up to `echo 5` + `F9` + `1` --> `echo 2` (it's 0-indexed!)

## F7
- Again, UI looks pretty different. This one is probably going to be the most noticeable one since it's widely used.
![UI comparison](https://github.com/microsoft/terminal/assets/11050425/227766e5-e0e8-4890-a46e-d3a20976ae67)
- ✅ given the new UI, if we don't have enough space at the bottom, we shift everything over
- ✅ if the retrieved command doesn't fit on the screen, we truncate it

Looks like there's just one bug (and it's probably fixed by a missed call to the renderer somewhere). Nice work!

Reply:
> Curious, why though? I'm guessing it's a limitation on switching over to just using VT?

Yes, but also not quite. We could implement the same popup interface as we had before, but it always had the problem that it reads the buffer contents. When you dismiss that old popup UI it needs to restore the text that was behind it after all. It's possible to make this Unicode correct, but it's also very annoying to do so.
However, there's another problem: When you restart a session ("press Enter to restart" in WT), it recreates a new ConPTY instance which does not have the previous buffer contents. When the popup reads the buffer contents it'll read whitespace and when you dismiss it, it'll destroy your previous buffer contents.
Both issues can be avoided by simply modernizing the popup UI a little, and by making sure that the "popup" is always below the prompt and never on top of it.

Reply:
The last 2 commits address the issue you found. You can pretty much ignore the "Fix buffer invalidation" one though. 😅 I've decided to go with a way simpler solution for now until we improve our grapheme cluster support even further at some point.

Reply:
Just on the subject of the popups restoring the text behind them, technically we should be able to achieve that now with VT pages, without having to read the buffer contents. The catch is that it's unlikely to work on most other terminals, so we'd still have to have a simpler fallback UI that doesn't rely on popups, and some form of feature detection at startup. I would have loved for us to do that, but I know that's asking a bit much.

Reply:
I really wouldn't mind adding something like that in the future! It sounds pretty cool and we could show off our more advanced VT features that way.